src/lib/blkalloc/append_blk_allocator.cpp

```diff
@@ -0,0 +1,180 @@
+/*********************************************************************************
+ * Modifications Copyright 2017-2019 eBay Inc.
+ *
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ * *
+ * *********************************************************************************/
+#include <homestore/meta_service.hpp>
+#include "append_blk_allocator.h"
+
+namespace homestore {
+
+AppendBlkAllocator::AppendBlkAllocator(const BlkAllocConfig& cfg, bool first_time_boot, chunk_num_t id) :
+        BlkAllocator{cfg, id}, m_metrics{get_name().c_str()}, m_sb{get_name()} {
+    // all append_blk_allocator instances use same client type;
+    meta_service().register_handler(
+        get_name(),
+        [this](meta_blk* mblk, sisl::byte_view buf, size_t size) { on_meta_blk_found(std::move(buf), (void*)mblk); },
+        nullptr);
+
+    if (first_time_boot) {
+        m_freeable_nblks = available_blks();
+        m_last_append_offset = 0;
+
+        m_sb.create(sizeof(append_blkalloc_sb));
+        m_sb->allocator_id = id;
+        m_sb->last_append_offset = 0;
+        m_sb->freeable_nblks = m_freeable_nblks;
+    }
+
+    // for recovery boot, fields should be recovered from metablks;
+}
+
+void AppendBlkAllocator::on_meta_blk_found(const sisl::byte_view& buf, void* meta_cookie) {
+    // TODO: also needs to initialize base class blkallocator for recovery path;
+
+    m_sb.load(buf, meta_cookie);
+    m_last_append_offset = m_sb->last_append_offset;
+    m_freeable_nblks = m_sb->freeable_nblks;
+
+    HS_REL_ASSERT_EQ(m_sb->magic, append_blkalloc_sb_magic, "Invalid AppendBlkAlloc metablk, magic mismatch");
+    HS_REL_ASSERT_EQ(m_sb->version, append_blkalloc_sb_version, "Invalid version of AppendBlkAllocator metablk");
+}
+
+// alloc a single block;
+BlkAllocStatus AppendBlkAllocator::alloc(BlkId& bid) {
+    if (available_blks() < 1) {
```

Comment:
it is possible for upper layer to call AppendBlkAllocator::alloc() at the same time? if yes, it seems it is not safely enough.  for example: there is only one last blk is left on this chunk and both two requests try to allocate one blk from this chunk and seems both requests will be successfully at last, but that is not expected.

Reply:
The assumption for append blk allocator is only one writer per each shard/chunk. 
We can choose to support multiple writers for appendblkallocator (if so, we should not be using append allocator, IMO), we can revisit this design and make required changes. Or do you see a need from your side we need to support multiple writers?

Reply:
there will be the case that s3 GW will send multi put blobs(from different objects ) request in one shard to homeobject, then homeobject will use call homestore data service alloc_blk() in parallel to alloc blk for writing. I think the case may happen, but we can avoid calling homestore data service alloc_blk() in parallel for a chunk by using lock in homeobject layer.

Reply:
one chunk one writer, I don't think HO will violate this rule.

Reply:
@zichanglai @yamingk Just wanted to make sure we are in same page: HO will not call alloc_blk or write directly. It is done by ReplDev. 

With that we have 2 choices:
1. Let ReplDev take a lock before every alloc_blk and surrounding async write
2. Handle this in AppendBlkAllocator and let it return safe blkids to allow concurrent allocation. But the subsequent writes are not locked and if it so happens that the writes come in out-of-order, so be it. Moreover, we might be doing buffered IO in HDD case, so its really not a problem.

I prefer Option: 2, because 
a) ReplDev is a common layer and not append_blk_allocator specific and we don't want to introduce logic.
b) We never take lock surrounding IO operations. This is all through the homestore and design goal/requirement.

Reply:
BTW: I am not suggesting to make this change as part of this PR, but please open up an issue to track this.

Reply:
ReplDev doesn't need to take lock surrounding async_write, concurrent write can still happen if they are not on same shard. 
If write arrives on same shard with multiple blobs, ReplDev only needs to make sure that those async_write on *same* shard happens in same thread (next async_write doesn't need to wait for previous async-write to complete). Again, no lock is needed.

Reply:
@yamingk ReplDev will not know anything about the shard, it will not ensure any per shard ordering. With the current flow after all the discussions, where HO doesn't participate in any write, ReplDev will call alloc_blks() to append_blk_allocator. Unless it takes a global lock, it is impossible to make it concurrent on each shards.

So the requirement to AppendBlkAllocator or any BlkAllocator for that matter is its allocation needs to be thread safe. I believe we can still achieve this atomics, but that is implementation details. 

I understand that this requirement has changed, since we discussed AppendBlkAllocator, so we may need to open an issue and resolve it in next PR.

Reply:
it is okay to change the requirement. 
Maybe I am not very clear on the IO threading model between HO and ReplDev. I thought HO will handle all the incoming I/Os and make sure I/Os sending to ReplDev in each thread won't clash (e.g. write on same shard will arrive on same thread), but I could be wrong.

Reply:
@hkadayam yes, I take the statement of HO will call alloc_blk() back, HO will not call alloc_blk() directly and ReplDev will do that. ReplDev is pg level and not aware of shard at all, unless holding a global lock, it is possible that concurrent writes will happen on the same shard, it will also happens on different shards too.
@yamingk I think we can try use CAS to achieve thread-safe alloc, for example:
std::atomic<uint64_t> last_offset;

BlkID alloc_blk(hints) {
       uint64_t now_offset = last_offset.load();
       do  {
            if ((get_total_blks() - now_offset) < 1) {
                  xxxxx 
                  return failed;
            }
            
            if (!last_offset.compare_exchange_strong(now_offset, now_offset + 1)) {
                  //race condition happens and now_offset is already be updated to the actually stored value in last_offset
                 continue;
            }

            //allocation is succeeded now;
            out_bids.emplace_back(now_offset, nblks, m_chunk_id);
           m_freeable_nblks -= nblks;
           set_dirty_flag();
           return. xxxx;

      } while (true);
}

but yes, even allocation is thread-safe and there is no guarantee that it will always be sequence IO while writing on disk. but I think it is already impossible for us to achieve that because many shards  on same disk are being writing concurrent.

Reply:
thread-safe and sequential I/O are likely exclusive, you need to give up one for the other. 
Thread-safe makes perfect sense for appendblkallocator to work with nvme drive. 
For HDD, I am trying to keep that option open (sequencial I/O will achieve better performance) open and leave consumer to decide. It is our implementation that choose to do concurrent writes on same shard/chunk, which is fine.

Reply:
closing this conversion with thread_safe mode added.